### PR TITLE
Backend/Fix/ Set DriverQuotes inactive on search cancel

### PR DESCRIPTION
**Pull request by DonDokidon**
## Type of Change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 1]
```diff
@@ -789,7 +789,9 @@ respondQuote (driverId, _) req = do
       case req.response of
         Pulled -> throwError UnexpectedResponseValue
         Accept -> do
-          when (searchReq.autoAssignEnabled == Just True) $ CS.incrementSearchRequestLockCounter searchTryId
+          when (searchReq.autoAssignEnabled == Just True) $
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:792]: this way how are we handling if any driver have already responded then cancel search should not work in autoassign flow.

Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:792]: > this way how are we handling if any driver have already responded then cancel search should not work in autoassign flow.

just checking the lock

Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:792]: @DonDokidon Previously, we were checking if driver has already given a quote then search cancel should fail or if search has cancelled and then driver gave the quote then driver quote should fail. 

But with your current implementation, if search has cancelled then driver can't give quote, or search can't be cancelled again.
But it didn't cover the case, when driver has given the quote then search cancelled shouldn't work.

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Comment by SrinivasNaikB** [comment]
linked [Issue](https://github.com/nammayatri/nammayatri/issues/1379)

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs (hunk 1 above)
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs]: > But with your current implementation, if search has cancelled then driver can't give quote, or search can't be cancelled again. But it didn't cover the case, when driver has given the quote then search cancelled shouldn't work.

Fixed

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 2]
```diff
@@ -827,7 +827,10 @@ respondQuote (driverId, _) req = do
       case req.response of
         Pulled -> throwError UnexpectedResponseValue
         Accept -> do
-          when (searchReq.autoAssignEnabled == Just True) $ CS.incrementSearchRequestLockCounter searchTryId
+          when (searchReq.autoAssignEnabled == Just True) $ do
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:830]: i can see race condition happening here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 3]
```diff
@@ -827,7 +827,10 @@ respondQuote (driverId, _) req = do
...
+            whenM (CS.isSearchTryCancelled searchTryId) $
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:831]: i can see race condition happening here.

Comment (piyushKumar-1) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:831]: both `cancelSearch` and `respondQuote` operations can be executed in this implementation.

Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:831]: > both `cancelSearch` and `respondQuote` operations can be executed in this implementation.

there was race condition before too. Ok, i'll fix it

Comment (piyushKumar-1) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:831]: ohh, i see. sure chances are very less, although if we can fix it properly that's better.

**Review by hkmangla — APPROVED** [review]

**Pull request merged**